tdms/src/interpolate_Hfield.cpp

```diff
@@ -1,161 +1,235 @@
         }
-        else
+        // now interpolate in the x-direction to the centre of Yee cell (i,j,k)
+        *Hy = x_scheme.interpolate(pass_to_second_scheme);
+    }
+    else
+    {
+        // we will be interpolating in the x-direction first, then in z
+        // this is the converse scenario described above
+        for (int kk = z_scheme.first_nonzero_coeff; kk <= z_scheme.last_nonzero_coeff; kk++)
         {
-            // we're using a cubic scheme in y, assign memory we need
-            other_dim_cell_data = (double *)malloc(4*sizeof(double));
-
-            // Each time we interpolate in y, we need interpolated data for cells j - (scheme-7+1) through j - (scheme-7+1) + 3, inclusive
-            d_low = j - (y_scheme - 6);
-            d_high = j - (y_scheme - 6) + 3;
-            for (int ii = i_low; ii <= i_high; ii++)
+            // this is the k-index of the cell we are looking at
+            int cell_k = k - (z_scheme.index + 1) + kk;
+            // determine the Hy values of cells (i - x_scheme.index-1, j, cell_k) through (i- x_scheme.index-1+7, j, cell_k), and interpolate them via x_scheme
+            for (int ii = x_scheme.first_nonzero_coeff; ii <= x_scheme.last_nonzero_coeff; ii++)
             {
-                // populate other_dim_cell_data with the data we need from the y-cells
-                for (int jj = d_low; jj <= d_high; jj++)
-                {
-                    other_dim_cell_data[jj - d_low] = Hxy[k][jj][ii] + Hxz[k][jj][ii];
-                }
-                first_interp_vals[ii - i_low] = cubic_interpolation(y_scheme-7, other_dim_cell_data);
+                // the i-index of the current cell we are looking at (readability, don't need to define this here)
+                int cell_i = i - (x_scheme.index + 1) + ii;
+                // gather the data for interpolating in the x dimension
+                data_for_first_scheme[ii] = Hyx[cell_k][j][cell_i] + Hyz[cell_k][j][cell_i];
             }
+            // interpolate in x to obtain a value for the Hy field at position (i, j, cell_k+Dz)
+            // place this into the appropriate index in the data being passed to the y_scheme
+            pass_to_second_scheme[kk] = x_scheme.interpolate(data_for_first_scheme);
         }
+        // now interpolate in the z-direction to the centre of Yee cell (i,j,k)
+        *Hy = z_scheme.interpolate(pass_to_second_scheme);
     }
-    else {
-        // prefer z-dimension over y-dimension
-        interp_scheme z_scheme = determineInterpScheme(K, k);
-        if (better_scheme(z_scheme, CUBIC_INTERP_MIDDLE))
-        {
-            // using a BLi scheme in z, assign memory we need
-            other_dim_cell_data = (double *)malloc(8 * sizeof(double));
+}
+
+void interpolateTimeDomainHz(double ***Hzx, double ***Hzy, int i, int j, int k, int I, int J, double *Hz)
+{
+    /* There is a choice in the order of interpolation; either interpolate in x first, then y, or vice-versa.
```

Comment:
this is a lot of comments to maintain. is there a way to write something generic suitable for all `interpolateTimeDomainH{x, y, z}` ?

Reply:
I'll break out the whole thing at the top of the file in a generic setting, then use shorter, more contextual inline comments in the functions themselves